avail/src/main/kotlin/avail/anvil/Styles.kt

```diff
@@ -32,605 +32,2987 @@
+			}
+			else
+			{
+				compile(
+					UnvalidatedStylePattern(
+						FailedMatchToken.lexeme,
+						UnvalidatedRenderingContext(StyleAttributes())
+					),
+					palette
+				)
+			}
+	}
+
+	/**
+	 * The root [tree][StyleRuleTree] for rendering queries.
+	 */
+	private val rootTree by lazy(LazyThreadSafetyMode.PUBLICATION) {
+		StyleRuleTree(rules.map { it.initialContext }.toSet())
+	}
+
+	/**
+	 * The cached [solutions][RenderingContext] for rendering queries. The input
+	 * text is a comma-separated list of style classifiers, such as
+	 * [styleToken][AvailLoader.styleToken] produces when classifying regions of
+	 * source code. This mechanism provides multiple options for the memory
+	 * manager to reduce memory pressure, e.g., dropping cache items, clearing
+	 * [references][SoftReference] to successor [trees][StyleRuleTree], etc.
+	 */
+	private val solutions = LRUCache(
+		softCapacity = 2000,
+		strongCapacity = 200,
+		transformer = ::computeRenderingContext
+	)
+
+	/**
+	 * Obtain the correct [rendering&#32;context][RenderingContext] for the
+	 * specified input text, using a cached solution if possible. The input text
+	 * is a comma-separated list of style classifiers, such as
+	 * [styleToken][AvailLoader.styleToken] produces when classifying regions of
+	 * source code.
+	 *
+	 * @param classifiersString
+	 *   The classification label of some region of source text, as a
+	 *   comma-separated list of style classifiers.
+	 * @return
+	 *   The [rendering&#32;context][RenderingContext] to use for the region of
+	 *   source text whence was abstracted the sequence of classifiers.
+	 */
+	operator fun get(classifiersString: String) = solutions[classifiersString]
+
+	/**
+	 * Compute the [rendering&#32;context][RenderingContext] for the specified
+	 * input text, which is a comma-separated list of style classifiers, such as
+	 * [styleToken][AvailLoader.styleToken] produces when classifying regions of
+	 * source code.
+	 *
+	 * @param classifiersString
+	 *   The classification label of some region of source text, as a
+	 *   comma-separated list of style classifiers.
+	 * @return
+	 *   The [rendering&#32;context][RenderingContext] to use for the target
+	 *   region of source text.
+	 */
+	private fun computeRenderingContext(
+		classifiersString: String
+	): ValidatedRenderingContext
+	{
+		val tree = findTree(classifiersString)
+		// We have reached the tree containing the solution set.
+		if (tree.solutions.isEmpty())
+		{
+			// There are no solutions, then apply the special rule for failed
+			// matches.
+			return noSolutionsRule.renderingContext
+		}
+		// Compute the solution frontier, i.e., the subset of maximally specific
+		// solutions from the solution set.
+		val solutions = mostSpecificSolutions(tree)
+		// Aggregate the final rendering context by combining the solutions in
+		// order, such that later solutions override earlier ones when aspects
+		// are in conflict.
+		return distillFinalSolution(solutions)
```

Comment:
Really? I might need some convincing on this one. I am unconvinced that nodes will tend to reuse the same final context sets. It's possible that it's true, but I'd like to hear a compelling argument before I add this change. Because if the assumption is wrong, then it will have a negligible impact on time and a negative impact on memory.